### PR TITLE
Helm: Allow users to view and use repositories from Helm Hub

### LINKED ISSUE #445: Helm: Allow users to view and use repositories from Helm Hub

### PR DESCRIPTION
**Pull request by richard-cox**
- fixes #445
- To review
  - Update register helm hub ui component

### CODE REVIEW

File: src/frontend/packages/core/src/features/endpoints/create-endpoint/create-endpoint-base-step/create-endpoint-base-step.component.ts [hunk 1]
```diff
@@ -83,13 +83,14 @@ export class CreateEndpointBaseStepComponent {
...
   }
-  constructor(public store: Store<GeneralEntityAppState>, ) {
+  constructor(public store: Store<GeneralEntityAppState>,) {
     // Need to filter the endpoint types on the tech preview flag
     this.tileSelectorConfig$ = store.select(selectSessionData()).pipe(
       first(),
       map(sessionData => {
         const techPreviewIsEnabled = sessionData.config.enableTechPreview || false;
         return entityCatalog.getAllEndpointTypes(techPreviewIsEnabled)
+          .filter(endpoint => !endpoint.definition.unRegisterable)
```
Comment (richard-cox) [inline src/frontend/packages/core/src/features/endpoints/create-endpoint/create-endpoint-base-step/create-endpoint-base-step.component.ts]: Singleton/count config

File: src/frontend/packages/store/src/entity-catalog/entity-catalog.types.ts [hunk 2]
```diff
@@ -206,7 +213,7 @@ export interface IStratosEntityBuilder<T extends IEntityMetadata, Y = any> {
   getLines?(): EntityRowBuilder<T>[];
   getSubTypeLabels?(entityMetadata: T): {
     singular: string,
-    plural: string
+    plural: string;
```
Comment (richard-cox) [inline src/frontend/packages/store/src/entity-catalog/entity-catalog.types.ts]: double check

File: src/frontend/packages/suse-extensions/src/custom/helm/helm-entity-factory.ts [hunk 3]
```diff
@@ -10,9 +10,11 @@ export const getMonocularChartId = (entity: MonocularChart) => entity.id;
 export const getHelmVersionId = (entity: HelmVersion) => entity.endpointId;
 
 export const HELM_ENDPOINT_TYPE = 'helm';
+export const HELM_REPO_ENDPOINT_TYPE = 'repo';
+export const HELM_HUB_ENDPOINT_TYPE = 'hub';
 
 const entityCache: {
-  [key: string]: EntitySchema
+  [key: string]: EntitySchema;
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/helm-entity-factory.ts]: double check

File: src/frontend/packages/suse-extensions/src/custom/helm/helm-hub-registration/helm-hub-registration.component.html [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<app-steppers cancel="/endpoints">
+  <app-step title="" [valid]="true" [onNext]="onNext" finishButtonText="Register" [valid]="noHelmHub$ | async">
+    <div>
+      <ng-container *ngIf="noHelmHub$ | async; else hasHelmHub">
+        <p>Helm Hub is a public repository for Helm Charts. </p>
+        <p>To browse and install these charts you need to register Helm Hub as an endpoint by clicking `Register` below.
+        </p>
+      </ng-container>
+      <ng-template #hasHelmHub>
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/helm-hub-registration/helm-hub-registration.component.html]: Remove

File: src/jetstream/plugins/monocular/main.go [hunk 5]
```diff
@@ -41,16 +48,23 @@ type Monocular struct {
 	devSyncPID      int
 }
 
+type HelmHubChart struct {
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go:52]: Check for monocular version. (utils responses, body api response)

File: src/jetstream/plugins/monocular/main.go [hunk 6]
```diff
@@ -135,26 +149,32 @@ func (m *Monocular) syncOnStartup() {
...
+			if ep.SubType == helmRepoEndpointType {
+				helmRepos = append(helmRepos, ep.Name)
+
+				// Is this an endpoint that we don't have charts for ?
+				if !arrayContainsString(repos, ep.Name) {
+					log.Infof("Syncing helm repository to chart store: %s", ep.Name)
+					m.Sync(interfaces.EndpointRegisterAction, ep)
+				}
+			} else {
 				log.Infof("Syncing helm repository to chart store: %s", ep.Name)
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go]: Remove

File: src/jetstream/plugins/monocular/main.go [hunk 7]
```diff
@@ -225,12 +245,45 @@ func (m *Monocular) AddSessionGroupRoutes(echoGroup *echo.Group) {
 	// API for Helm Chart Repositories
 	echoGroup.GET("/chartrepos", m.ListRepos)
 	echoGroup.Any("/chartsvc/*", m.handleAPI)
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go:243]: move chartsvc and monocular to top, only ones used for monocular

File: src/jetstream/plugins/monocular/main.go [hunk 8]
```diff
@@ -225,12 +245,45 @@ func (m *Monocular) AddSessionGroupRoutes(echoGroup *echo.Group) {
...
+	if cnsiList == nil || len(cnsiList) == 0 || len(cnsiList[0]) == 0 {
+		return nil, nil
+	}
+
+	return m.validateMonocularInstanceRequest(cnsiList[0])
+}
+
+func (m *Monocular) validateMonocularInstanceRequest(cnsi string) (*interfaces.CNSIRecord, error) {
+	endpoint, err := m.portalProxy.GetCNSIRecord(cnsi)
+	if err != nil {
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go:269]: check for helm type

File: src/jetstream/plugins/monocular/main.go [hunk 9]
```diff
@@ -225,12 +245,45 @@ func (m *Monocular) AddSessionGroupRoutes(echoGroup *echo.Group) {
...
 	echoGroup.POST("/chartrepos/status", m.GetRepoStatuses)
 	echoGroup.POST("/chartrepos/:guid", m.SyncRepo)
 }
 
+// Forward requests to the Chart Service API
+func (m *Monocular) isMonocularInstanceRequest(c echo.Context) (*interfaces.CNSIRecord, error) {
+	cnsiList := strings.Split(c.Request().Header.Get("x-cap-cnsi-list"), ",")
+
+	// If this has a cnsi then treat it as a monocular request
+	if cnsiList == nil || len(cnsiList) == 0 || len(cnsiList[0]) == 0 {
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go]: check for exactly one

File: src/jetstream/plugins/monocular/main.go [hunk 10]
```diff
@@ -225,12 +245,45 @@ func (m *Monocular) AddSessionGroupRoutes(echoGroup *echo.Group) {
...
+		err := errors.New("Failed to fetch endpoint")
+		return nil, echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	}
+
+	return &endpoint, nil
+}
+
 // Forward requests to the Chart Service API
 func (m *Monocular) handleAPI(c echo.Context) error {
+	monocularEndpoint, err := m.isMonocularInstanceRequest(c)
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go]: isNotBuiltInMonocularRequest

File: src/jetstream/plugins/monocular/main.go [hunk 11]
```diff
@@ -225,12 +245,45 @@ func (m *Monocular) AddSessionGroupRoutes(echoGroup *echo.Group) {
...
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	}
+	if monocularEndpoint != nil {
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go]: add comment, not built in monocular forwarding on to remote

File: src/jetstream/plugins/monocular/main.go [hunk 12]
```diff
@@ -242,3 +295,145 @@ func (m *Monocular) handleAPI(c echo.Context) error {
...
+
+		// Forwarding everything else
+		default:
+			emptyRequest.Header[k] = v
+		}
+	}
+}
+
+// Forward request to monocular of endpoint
+func (m *Monocular) baseHandleMonocularInstance(c echo.Context, monocularEndpoint *interfaces.CNSIRecord) error {
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go:337]: should be able to use doProxy, create issue. include issue about auth

File: src/jetstream/plugins/monocular/main.go [hunk 13]
```diff
@@ -242,3 +295,145 @@ func (m *Monocular) handleAPI(c echo.Context) error {
...
+		c.Response().Write([]byte(err.Error()))
+	} else if res.Body != nil {
+		c.Response().Status = res.StatusCode
+		c.Response().Header().Set("Content-Type", res.Header.Get("Content-Type"))
+		body, _ := ioutil.ReadAll(res.Body)
+		c.Response().Write(body)
+		defer res.Body.Close()
+	} else {
+		c.Response().Status = 200
+		c.Response().Write([]byte("{ data: [] }"))
```
Comment (richard-cox) [inline src/jetstream/plugins/monocular/main.go]: Remove

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular.interceptor.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    if (!this.includeUrls.find(part => req.url.indexOf(part) >= 0)) {
+      return next.handle(req);
+    }
+
+    const endpoint = this.route.snapshot.params.endpoint;
+    if (!endpoint) {
+      return next.handle(req);
+    }
+
+    if (endpoint === stratosMonocularEndpointGuid) {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular.interceptor.ts]: tidy logic

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular/chart-item/chart-item.component.ts [hunk 15]
```diff
@@ -26,11 +27,11 @@ export class ChartItemComponent implements OnInit {
...
 
   goToDetailUrl(): string {
-    return `/monocular/charts/${this.chart.attributes.repo.name}/${this.chart.attributes
+    return `/monocular/charts/${getMonocularEndpoint(null, this.chart)}/${this.chart.attributes.repo.name}/${this.chart.attributes
       .name}`;
   }
 
   goToRepoUrl(): string {
-    return `/monocular/charts/${this.chart.attributes.repo.name}`;
+    return `/monocular/charts//${getMonocularEndpoint(null, this.chart)}/${this.chart.attributes.repo.name}`;
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular/chart-item/chart-item.component.ts]: //

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts [hunk 16]
```diff
@@ -48,7 +70,7 @@ export class ChartsService {
         observer.next(this.cacheCharts[repo]);
       });
     } else {
-      return this.http.get<{ data: any }>(url).pipe(
+      return this.http.get<{ data: any; }>(url).pipe(
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts]: double-check

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts [hunk 17]
```diff
@@ -19,13 +21,33 @@ export class ChartsService {
...
+   * Update url to go to a monocular instance other than stratos
+   * These are used as img src values so won't hit our http interceptor
+   */
+  private updateStratosUrl(chart: Chart, url: string): string {
+    // former if chart came from charts request, latter if chart came directly from monocular
+    const endpoint = chart.stratosEndpointId || this.route.snapshot.params.endpoint;
+    if (!endpoint || endpoint === stratosMonocularEndpointGuid) {
+      return url;
+    }
+    // "/pp/v1/chartsvc" "/v1/assets/billimek/adguard-home/logo"
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts]: Remove

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/repos.service.ts [hunk 18]
```diff
@@ -5,8 +5,6 @@ import { catchError, map } from 'rxjs/operators';
 
 import { LoggerService } from '../../../../../../../core/src/core/logger.service';
 import { RepoAttributes } from '../models/repo';
-import { ConfigService } from './config.service';
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/repos.service.ts:8]: add back in

File: src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts [hunk 19]
```diff
@@ -19,13 +21,33 @@ export class ChartsService {
...
     this.hostname = '/pp/v1/chartsvc';
   }
 
+  /**
+   * Update url to go to a monocular instance other than stratos
+   * These are used as img src values so won't hit our http interceptor
+   */
+  private updateStratosUrl(chart: Chart, url: string): string {
+    // former if chart came from charts request, latter if chart came directly from monocular
+    const endpoint = chart.stratosEndpointId || this.route.snapshot.params.endpoint;
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/monocular/shared/services/charts.service.ts]: can use getMonocularEndpoint?

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.effects.ts [hunk 20]
```diff
@@ -76,25 +130,41 @@ export class HelmEffects {
...
-          return res;
-        }, base);
-        return processedData;
-      }, []);
+      this.store.dispatch(new StartRequestAction(action));
+
+      const helmEndpoints = Object.values(endpointOfTypeSelector(HELM_ENDPOINT_TYPE)(appState));
+      const helmHubEndpoint = helmEndpoints.find(endpoint => endpoint.sub_type === HELM_HUB_ENDPOINT_TYPE);
+
+      const stratosMonocular = this.httpClient.get<MonocularChartsResponse>(`/pp/${this.proxyAPIVersion}/chartsvc/v1/charts`);
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/store/helm.effects.ts:139]: Create Issue - investigate a way to use the cnis-list header to go out to all helm types

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.effects.ts [hunk 21]
```diff
@@ -203,7 +294,7 @@ export class HelmEffects {
     return 'Helm API request error';
   }
 
-  static createHelmError(err: any): { status: string, message: string } {
+  static createHelmError(err: any): { status: string, message: string; } {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/store/helm.effects.ts]: double check

File: src/jetstream/plugins/metrics/main.go [hunk 22]
```diff
@@ -348,9 +348,6 @@ func (m *MetricsSpecification) UpdateMetadata(info *interfaces.Info, userGUID st
 	for _, values := range info.Endpoints {
 		for _, endpoint := range values {
 			// Look to see if we can find the metrics provider for this URL
-			log.Debugf("Processing endpoint: %+v", endpoint)
-			log.Debugf("Processing endpoint: %+v", endpoint.CNSIRecord)
```
Comment (richard-cox) [inline src/jetstream/plugins/metrics/main.go:352]: keep this line

**Comment by richard-cox** [comment]
I've updated following review comments. The Charts list repo filter --> side bar change is still pending

File: src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts [hunk 23]
```diff
@@ -51,14 +51,20 @@ export interface HelmInstallValues {
   releaseName: string;
   releaseNamespace: string;
   values: string;
-  chart: string;
+  chart: {
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/store/helm.types.ts:54]: find common type (name/repo/version)

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts [hunk 24]
```diff
@@ -33,15 +34,19 @@ import { ReleaseUpgradeVersionsListConfig } from './release-version-list-config'
...
   public overrides: FormGroup;
 
   // Future
   public showAdvancedOptions = false;
 
-  constructor(store: Store<any>, public helper: HelmReleaseHelperService) {
+  constructor(
+    store: Store<any>,
+    public helper: HelmReleaseHelperService,
+    private route: ActivatedRoute,
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts]: route can go?

File: src/frontend/packages/suse-extensions/src/custom/helm/create-release/create-release.component.ts [hunk 25]
```diff
@@ -256,7 +256,7 @@ export class CreateReleaseComponent implements OnInit, OnDestroy {
       ...this.details.value,
       ...this.overrides.value,
       chart: {
-        name: this.route.snapshot.params.name,
+        chartName: this.route.snapshot.params.chartName,
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/helm/create-release/create-release.component.ts:259]: change to name

File: src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts [hunk 26]
```diff
@@ -62,7 +63,8 @@ export class UpgradeReleaseComponent {
       const name = chart.upgrade.name;
       const repoName = chart.upgrade.repo.name;
       const version = chart.release.chart.metadata.version;
-      this.listConfig = new ReleaseUpgradeVersionsListConfig(store, repoName, name, version);
+      this.listConfig = new ReleaseUpgradeVersionsListConfig(store, repoName, name, version, chart.monocularEndpointId);
```
Comment (richard-cox) [inline src/frontend/packages/suse-extensions/src/custom/kubernetes/workloads/upgrade-release/upgrade-release.component.ts:66]: const version = chart.release.chart.metadata.version; pattern

**Review by nwmac — APPROVED** [review]
LGTM

**Pull request merged**